SwedishStemmer (and DutchStemmer?) not thread safe

While using a Swedish language Maui Server project concurrently from multiple processes, I got several 500 Internal Server Errors with the following traceback:

```
26 May 2020 10:15:37 ERROR RootServlet - Uncaught exception in servlet
org.topbraid.mauiserver.MauiServerException: Error running recommender: String index out of range: 8
        at org.topbraid.mauiserver.tagger.Tagger.recommendTags(Tagger.java:141)
        at org.topbraid.mauiserver.SuggestResource.doSuggest(SuggestResource.java:54)
        at org.topbraid.mauiserver.SuggestResource.doPost(SuggestResource.java:46)
        at org.topbraid.mauiserver.framework.RootServlet.createResponse(RootServlet.java:111)
        at org.topbraid.mauiserver.framework.RootServlet.doRequest(RootServlet.java:65)
        at org.topbraid.mauiserver.framework.RootServlet.doPost(RootServlet.java:42)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:648)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:729)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:292)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:207)
        at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:52)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:240)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:207)
        at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:213)
        at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:106)
        at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:494)
        at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:141)
        at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:79)
        at org.apache.catalina.valves.AbstractAccessLogValve.invoke(AbstractAccessLogValve.java:616)
        at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:88)
        at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:522)
        at org.apache.coyote.http11.AbstractHttp11Processor.process(AbstractHttp11Processor.java:1095)
        at org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:672)
        at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1520)
        at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.run(NioEndpoint.java:1476)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
        at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.StringIndexOutOfBoundsException: String index out of range: 8
        at java.lang.StringBuffer.charAt(StringBuffer.java:202)
        at org.tartarus.snowball.SnowballProgram.find_among_b(SnowballProgram.java:278)
        at org.tartarus.snowball.ext.swedishStemmer.r_main_suffix(swedishStemmer.java:178)
        at org.tartarus.snowball.ext.swedishStemmer.stem(swedishStemmer.java:325)
        at com.entopix.maui.stemmers.SwedishStemmer.stem(SwedishStemmer.java:16)
        at com.entopix.maui.vocab.Vocabulary.pseudoPhrase(Vocabulary.java:657)
        at com.entopix.maui.vocab.Vocabulary.normalizePhrase(Vocabulary.java:622)
        at com.entopix.maui.vocab.Vocabulary.getSenses(Vocabulary.java:521)
        at com.entopix.maui.filters.MauiFilter.getCandidates(MauiFilter.java:1330)
        at com.entopix.maui.filters.MauiFilter.convertInstance(MauiFilter.java:1041)
        at com.entopix.maui.filters.MauiFilter.input(MauiFilter.java:541)
        at com.entopix.maui.main.MauiWrapper.extractTopicsFromText(MauiWrapper.java:163)
        at org.topbraid.mauiserver.tagger.Tagger.recommendTags(Tagger.java:133)
        ... 28 more
```

The root cause seems to be that the Snowball stemmer used by SwedishStemmer is not thread safe. (see [a similar issue](https://github.com/eclipse/deeplearning4j/issues/31) in another project)
I think the problem is that SwedishStemmer in maui stores a single stemmer and this breaks when accessed from multiple servlet threads in Maui Server.

Since the new DutchStemmer, just added in PR #9, uses a similar Snowball stemmer it most likely suffers from the same problem, although I haven't tested it.

--- a/src/main/java/com/entopix/maui/stemmers/DutchStemmer.java
+++ b/src/main/java/com/entopix/maui/stemmers/DutchStemmer.java
@@ -5,13 +5,9 @@ import org.tartarus.snowball.ext.dutchStemmer;
 public class DutchStemmer extends Stemmer {
 	private static final long serialVersionUID = 1L;
 
-	private transient dutchStemmer snowball;
-
 	@Override
 	public String stem(String str) {
-		if (snowball == null) {
-			snowball = new dutchStemmer();
-		}
+		dutchStemmer snowball = new dutchStemmer();
 		snowball.setCurrent(str);
 		snowball.stem();
 		return snowball.getCurrent();

--- a/src/main/java/com/entopix/maui/stemmers/SwedishStemmer.java
+++ b/src/main/java/com/entopix/maui/stemmers/SwedishStemmer.java
@@ -5,13 +5,9 @@ import org.tartarus.snowball.ext.swedishStemmer;
 public class SwedishStemmer extends Stemmer {
 	private static final long serialVersionUID = 1L;
 	
-	private transient swedishStemmer snowball;
-	
 	@Override
 	public String stem(String str) {
-		if (snowball == null) {
-			snowball = new swedishStemmer();
-		}
+		swedishStemmer snowball = new swedishStemmer();
 		snowball.setCurrent(str);
 		snowball.stem();
 		return snowball.getCurrent();

--- a/src/test/java/com/entopix/maui/main/DutchStemmerTest.java
+++ b/src/test/java/com/entopix/maui/main/DutchStemmerTest.java
@@ -1,0 +1,26 @@
+package com.entopix.maui.main;
+
+import static org.junit.Assert.*;
+
+import org.junit.Before;
+import org.junit.Test;
+
+import com.entopix.maui.stemmers.DutchStemmer;
+import com.entopix.maui.stemmers.Stemmer;
+
+public class DutchStemmerTest {
+
+	private Stemmer stemmer;
+	
+	@Before
+	public void setUp() throws Exception {
+		stemmer = new DutchStemmer();
+	}
+
+	@Test
+	public void testHuis() {
+		assertEquals("huis", stemmer.stem("huis"));
+		assertEquals("huiz", stemmer.stem("huizen")); // the result is not ideal but that's what Snowball does
+	}
+
+}

--- a/src/test/java/com/entopix/maui/main/SwedishStemmerTest.java
+++ b/src/test/java/com/entopix/maui/main/SwedishStemmerTest.java
@@ -1,0 +1,28 @@
+package com.entopix.maui.main;
+
+import static org.junit.Assert.*;
+
+import org.junit.Before;
+import org.junit.Test;
+
+import com.entopix.maui.stemmers.SwedishStemmer;
+import com.entopix.maui.stemmers.Stemmer;
+
+public class SwedishStemmerTest {
+
+	private Stemmer stemmer;
+	
+	@Before
+	public void setUp() throws Exception {
+		stemmer = new SwedishStemmer();
+	}
+
+	@Test
+	public void testKatt() {
+		assertEquals("katt", stemmer.stem("katt"));
+		assertEquals("katt", stemmer.stem("katten"));
+		assertEquals("katt", stemmer.stem("katter"));
+		assertEquals("katt", stemmer.stem("katterna"));
+	}
+
+}

